File: app/src/main/java/org/deiverbum/app/model/Today.java
package org.deiverbum.app.model;

import android.text.SpannableStringBuilder;

import androidx.room.Ignore;

import org.deiverbum.app.utils.Utils;

/**
 * <p>
 *     Esta clase recoge información valiosa sobre el día litúrgico.
 * </p>
 * @author A. Cedano
 * @version 1.0
 * @since 2022.1
 */

//@SuppressWarnings("SameReturnValue")
public class Today {
    public Integer todayDate;
    public Integer weekDay;
    public Integer timeID;
    @Ignore
    public Integer weekDayFK;
    public Integer liturgyFK;
    public Integer previousFK;
    public Integer massReadingFK;
    public Integer invitatoryFK;

    //@Ignore
    public Integer hasSaint;
    public Integer saintFK;


    public Integer oHymnFK;
    public Integer oPsalmodyFK;
    public Integer oVerseFK;
    public Integer oBiblicalFK;
    public Integer oPatristicFK;
    public Integer oPrayerFK;
    public Integer oTeDeum;

    public Integer lHymnFK;
    public Integer lPsalmodyFK;
    public Integer lBiblicalFK;
    public Integer lBenedictusFK;
    public Integer lIntercessionsFK;
    public Integer lPrayerFK;

    public Integer tHymnFK;
    public Integer tPsalmodyFK;
    public Integer tBiblicalFK;
    public Integer tPrayerFK;

    public Integer sHymnFK;
    public Integer sPsalmodyFK;
    public Integer sBiblicalFK;
    public Integer sPrayerFK;

    public Integer nHymnFK;
    public Integer nPsalmodyFK;
    public Integer nBiblicalFK;
    public Integer nPrayerFK;

    public Integer vHymnFK;
    public Integer vPsalmodyFK;
    public Integer vBiblicalFK;
    public Integer vMagnificatFK;
    public Integer vIntercessionsFK;
    public Integer vPrayerFK;

    @Ignore
    public Liturgy liturgyDay;
    @Ignore
    public Liturgy liturgyPrevious;
    @Ignore
    public LiturgyTime liturgyTime;

    public Today() {
    }

    public Integer getTodayDate() {
        return todayDate;
    }

    public void setTodayDate(Integer todayDate) {
        this.todayDate = todayDate;
    }

    public Integer getMassReadingFK() {
        return massReadingFK;
    }

    public void setMassReadingFK(Integer massReadingFK) {
        this.massReadingFK = massReadingFK;
    }

    public Integer getPreviousFK() {
        return previousFK;
    }

    public void setPreviousFK(Integer previousFK) {
        this.previousFK = previousFK;
    }

    public Integer getTimeID() {
        return timeID;
    }

    public void setTimeID(Integer timeID) {
        this.timeID = timeID;
    }

    public void setHasSaint(int hasSaint) {
        this.hasSaint = hasSaint;
    }

    public Integer getHoy() {
        return todayDate;
    }

    public void setHoy(Integer hoy) {
        this.todayDate = hoy;
    }

    public void setMLecturasFK(Integer mLecturasFK) {
        this.massReadingFK = mLecturasFK;
    }

    public Integer getPrevioId() {
        return previousFK;
    }

    public void setPrevioId(Integer previo_id) {
        this.previousFK = previo_id;
    }

    public Integer getTiempoId() {
        return timeID;
    }

    public void setTiempoId(Integer tiempo_id) {
        this.timeID = tiempo_id;
    }

    public String getTituloVisperas() {
        if (liturgyPrevious != null) {
            return liturgyPrevious.getNombre().replaceAll(" I Vísperas.| I Vísperas","");
        } else {
            return liturgyDay.getNombre();
        }
    }



    public String getTitulo() {
            return (liturgyDay.typeID==6) ?
                    getTituloVisperas(): liturgyDay.getName();
    }

    public String getTituloForRead() {
        return (liturgyDay.typeID==6) ?
                getTituloVisperas(): liturgyDay.getTitleForRead();
    }

    public String getFecha() {
        return (todayDate != null) ? Utils.getLongDate(String.valueOf(todayDate)) : "";
    }

    public String getTiempo() {
        return (liturgyDay.typeID==6 && liturgyPrevious!=null) ?
                liturgyPrevious.getLiturgiaTiempo().getLiturgyName(): liturgyDay.getLiturgiaTiempo().getLiturgyName();
    }

    public void setWeekDay(int weekDay) {
        this.weekDay = weekDay;
    }

    public int getWeekDay() {
        return (this.weekDay!=null) ? weekDay : 0;
    }

    public int getHasSaint() {
        return this.hasSaint;
    }

    public boolean hasSaintToday() {
        return this.hasSaint == 1;
    }


    public SpannableStringBuilder getAllForView(boolean hasInvitatory) {
        liturgyDay.setHasSaint(hasSaintToday());
        SpannableStringBuilder sb = new SpannableStringBuilder();
        try {
            sb.append(Utils.LS);
            sb.append(getFecha());
            sb.append(Utils.LS2);
            sb.append(Utils.toH2(getTiempo()));
            sb.append(Utils.LS2);
            sb.append(Utils.toH3(getTitulo()));
            //liturgyDay.today.previousFK
            //sb.append(liturgyDay.getForView(hasInvitatory,previousFK));
            if (liturgyDay.typeID == 0) {
                if (oBiblicalFK==600010101){
                    sb.append(liturgyDay.getBreviaryHour().getOficioEaster().getForView());
                } else {
                    sb.append(liturgyDay.getBreviaryHour().getMixtoForView(liturgyDay.liturgyTime));//.getForView(liturgyDay.liturgyTime));
                }
            }
            if (liturgyDay.typeID == 1) {
                if (oBiblicalFK==600010101){
                    sb.append(liturgyDay.getBreviaryHour().getOficioEaster().getForView());
                }else {
                    sb.append(liturgyDay.getBreviaryHour().getOficio(hasInvitatory).getForView(liturgyDay.liturgyTime, hasSaintToday()));
                }
            }
            if (liturgyDay.typeID == 2) {
                sb.append(liturgyDay.getBreviaryHour().getLaudes(hasInvitatory).getForView(liturgyDay.liturgyTime));
            }
            if (liturgyDay.typeID == 3 || liturgyDay.typeID == 4 || liturgyDay.typeID == 5) {
                sb.append(liturgyDay.getBreviaryHour().getIntermedia().getForView(liturgyDay.liturgyTime,liturgyDay.typeID));
            }
            if (liturgyDay.typeID == 6) {
                sb.append(liturgyDay.getBreviaryHour().getVisperas().getForView(liturgyDay.liturgyTime));
            }
            if (liturgyDay.typeID == 7) {
                sb.append(liturgyDay.getBreviaryHour().getCompletas().getAllForView(liturgyDay.liturgyTime));
            }

        } catch (Exception e) {
            sb.append(Utils.createErrorMessage(e.getMessage()));
        }
        return sb;
    }

    public SpannableStringBuilder getSingleForView() {
        SpannableStringBuilder sb = new SpannableStringBuilder();
        try {
            sb.append(Utils.LS);
            sb.append(getFecha());
            sb.append(Utils.LS2);
            sb.append(Utils.toH2(getTiempo()));
            sb.append(Utils.LS2);
            sb.append(Utils.toH3(getTitulo()));
            //liturgyDay.today.previousFK
            //sb.append(liturgyDay.getAllForView());
        } catch (Exception e) {
            sb.append(Utils.createErrorMessage(e.getMessage()));
        }
        return sb;
    }

    public StringBuilder getAllForRead(boolean hasInvitatory) {
        StringBuilder sb = new StringBuilder();
        try {
            sb.append(Utils.pointAtEnd(getFecha()));
            sb.append(getTituloForRead());
            //sb.append(liturgyDay.getForRead());
            if (liturgyDay.typeID == 0) {
                if (oBiblicalFK==600010101){
                    sb.append(liturgyDay.getBreviaryHour().getOficioEaster().getForRead());
                } else {
                    //TODO
                    sb.append(liturgyDay.getBreviaryHour().getMixtoForRead());
                }
            }
            if (liturgyDay.typeID == 1) {
                if (oBiblicalFK==600010101){
                    sb.append(liturgyDay.getBreviaryHour().getOficioEaster().getForRead());
                }else {
                    sb.append(liturgyDay.getBreviaryHour().getOficio(hasInvitatory).getForRead());
                }
            }
            if (liturgyDay.typeID == 2) {
                sb.append(liturgyDay.getBreviaryHour().getLaudes(hasInvitatory).getForRead());
            }
            if (liturgyDay.typeID == 3 || liturgyDay.typeID == 4 || liturgyDay.typeID == 5) {
                sb.append(liturgyDay.getBreviaryHour().getIntermedia().getForRead());
            }
            if (liturgyDay.typeID == 6) {
                sb.append(liturgyDay.getBreviaryHour().getVisperas().getAllForRead(previousFK));
            }
            if (liturgyDay.typeID == 7) {
                sb.append(liturgyDay.getBreviaryHour().getCompletas().getAllForView(liturgyDay.liturgyTime));
            }

        } catch (Exception e) {
            sb.append(Utils.createErrorMessage(e.getMessage()));
        }
        return sb;
    }

    public StringBuilder getSingleForRead() {
        StringBuilder sb = new StringBuilder();
        try {
            sb.append(Utils.pointAtEnd(getFecha()));
            sb.append(getTituloForRead());
            //sb.append(liturgyDay.getForRead());

        } catch (Exception e) {
            sb.append(Utils.createErrorMessage(e.getMessage()));
        }
        return sb;
    }
}

File: app/src/main/java/org/deiverbum/app/model/Today.kt
package org.deiverbum.app.model

import android.text.SpannableStringBuilder
import androidx.room.Ignore
import org.deiverbum.app.domain.model.TodayRequest
import org.deiverbum.app.util.ColorUtils
import org.deiverbum.app.util.Utils

/**
 *
 *
 * Esta clase recoge información valiosa sobre el día litúrgico.
 *
 *
 * @author A. Cedano
 * @version 1.0
 * @since 2022.1
 */

class Today {
    var todayDate: Int = 0

    var weekDay: Int? = null
        get() = if(field==null) 0 else field

    var timeID: Int? = null

    @Ignore
    var weekDayFK: Int? = null
    var liturgyFK: Int? = null
    var previousFK: Int? = null
    var massReadingFK: Int? = null
    var invitatoryFK: Int? = null

    //@Ignore
    var hasSaint: Int? = null
    @JvmField
    var saintFK: Int? = null
    var oHymnFK: Int? = null
    var oPsalmodyFK: Int? = null
    var oVerseFK: Int? = null
    @JvmField
    var oBiblicalFK: Int? = null
    var oPatristicFK: Int? = null
    var oPrayerFK: Int? = null
    var oTeDeum: Int? = null
    var lHymnFK: Int? = null
    var lPsalmodyFK: Int? = null
    var lBiblicalFK: Int? = null
    var lBenedictusFK: Int? = null
    var lIntercessionsFK: Int? = null
    var lPrayerFK: Int? = null
    var tHymnFK: Int? = null
    var tPsalmodyFK: Int? = null
    var tBiblicalFK: Int? = null
    var tPrayerFK: Int? = null
    var sHymnFK: Int? = null
    var sPsalmodyFK: Int? = null
    var sBiblicalFK: Int? = null
    var sPrayerFK: Int? = null
    var nHymnFK: Int? = null
    var nPsalmodyFK: Int? = null
    var nBiblicalFK: Int? = null
    var nPrayerFK: Int? = null
    var vHymnFK: Int? = null
    var vPsalmodyFK: Int? = null
    var vBiblicalFK: Int? = null
    var vMagnificatFK: Int? = null
    var vIntercessionsFK: Int? = null
    var vPrayerFK: Int? = null
    var nightPrayerFK: Int? = null

    @JvmField
    @Ignore
    var liturgyDay: Liturgy? = null

    @JvmField
    @Ignore
    var liturgyPrevious: Liturgy? = null

    @Ignore
    var liturgyTime: LiturgyTime? = null
    fun setMLecturasFK(mLecturasFK: Int?) {
        massReadingFK = mLecturasFK
    }

    private val tituloVisperas: String
        get() = if (liturgyPrevious != null) {
            liturgyPrevious!!.name.replace(" I Vísperas.| I Vísperas".toRegex(), "")
        } else {
            liturgyDay!!.name
        }
    val titulo: String
        get() = if (liturgyDay!!.typeID == 6) tituloVisperas else liturgyDay!!.name
    private val tituloForRead: String
        get() = if (liturgyDay!!.typeID == 6) tituloVisperas else liturgyDay!!.titleForRead
    val fecha: String
        get() = Utils.getLongDate(todayDate.toString())
    val tiempo: String?
        get() = if (liturgyDay!!.typeID == 6 && liturgyPrevious != null) liturgyPrevious!!.liturgyTime!!.liturgyName else liturgyDay!!.liturgyTime!!.liturgyName


    private fun hasSaintToday(): Boolean {
        return hasSaint != null && hasSaint == 1
    }

     fun getAllForView(todayRequest:TodayRequest): SpannableStringBuilder {
        liturgyDay!!.setHasSaint(hasSaintToday())
        ColorUtils.isNightMode = todayRequest.isNightMode
        val sb = SpannableStringBuilder()
        try {
            sb.append(fecha)
            sb.append(Utils.LS2)
            sb.append(Utils.toH2(tiempo))
            sb.append(Utils.LS2)
            sb.append(Utils.toH3(titulo))

            if (liturgyDay!!.typeID == 9) {
                sb.append(liturgyDay?.homilyList?.getAllForView(todayRequest)
                )
                return sb
            }

            if (liturgyDay!!.typeID == 10) {
                sb.append(liturgyDay!!.massReadingList?.getForView(
                    todayRequest
                )
                )
            }

            val bh: BreviaryHour? =liturgyDay!!.breviaryHour

            if (liturgyDay!!.typeID == 0) {
                if (oBiblicalFK == 600010101) {
                    sb.append(bh?.getOficioEaster()?.forView)
                } else {
                    sb.append(
                        bh?.getMixtoForView(
                            liturgyDay!!.liturgyTime!!,
                            hasSaintToday()
                        )
                    )
                }
            }
            if (liturgyDay!!.typeID == 1) {
                if (oBiblicalFK == 600010101) {
                    sb.append(liturgyDay!!.breviaryHour!!.getOficioEaster()?.forView)
                } else {
                    sb.append(
                        liturgyDay!!.breviaryHour!!.getOficio(todayRequest.isMultipleInvitatory)!!
                            .getForView(liturgyDay!!.liturgyTime, hasSaintToday())
                    )
                }
            }
            if (liturgyDay!!.typeID == 2) {
                sb.append(
                    liturgyDay!!.breviaryHour!!.getLaudes(todayRequest.isMultipleInvitatory)!!
                        .getForView(liturgyDay!!.liturgyTime!!, hasSaintToday())
                )
            }
            if (liturgyDay!!.typeID == 3 || liturgyDay!!.typeID == 4 || liturgyDay!!.typeID == 5) {
                sb.append(
                    liturgyDay!!.breviaryHour!!.getIntermedia()!!
                        .getForView(liturgyDay!!.liturgyTime!!, liturgyDay!!.typeID)
                )
            }
            if (liturgyDay!!.typeID == 6) {
                sb.append(
                    liturgyDay!!.breviaryHour!!.getVisperas()!!
                        .getForView(liturgyDay!!.liturgyTime)
                )
            }
            if (liturgyDay!!.typeID == 7) {
                sb.append(liturgyDay!!.breviaryHour!!.getCompletas()!!.getAllForView())
            }
        } catch (e: Exception) {
            sb.append(Utils.createErrorMessage(e.message))
        }
        return sb
    }

    val singleForView: SpannableStringBuilder
        get() {
            val sb = SpannableStringBuilder()
            try {
                sb.append(Utils.LS)
                sb.append(fecha)
                sb.append(Utils.LS2)
                sb.append(Utils.toH2(tiempo))
                sb.append(Utils.LS2)
                sb.append(Utils.toH3(titulo))
            } catch (e: Exception) {
                sb.append(Utils.createErrorMessage(e.message))
            }
            return sb
        }

    fun getAllForRead(hasInvitatory: Boolean): StringBuilder {
        val sb = StringBuilder()
        try {
            sb.append(Utils.pointAtEnd(fecha))
            sb.append(tituloForRead)

            if (liturgyDay!!.typeID == 9) {
                sb.append(liturgyDay?.homilyList?.allForRead
                )
                return sb
            }
            if (liturgyDay!!.typeID == 0) {
                if (oBiblicalFK == 600010101) {
                    sb.append(liturgyDay!!.breviaryHour!!.getOficioEaster()?.forRead)
                } else {
                    //TODO
                    sb.append(liturgyDay!!.breviaryHour!!.getMixtoForRead())
                }
            }
            if (liturgyDay!!.typeID == 1) {
                if (oBiblicalFK == 600010101) {
                    sb.append(liturgyDay!!.breviaryHour!!.getOficioEaster()?.forRead)
                } else {
                    sb.append(liturgyDay!!.breviaryHour!!.getOficio(hasInvitatory)?.forRead)
                }
            }
            if (liturgyDay!!.typeID == 2) {
                sb.append(liturgyDay!!.breviaryHour!!.getLaudes(hasInvitatory)!!.forRead)
            }
            if (liturgyDay!!.typeID == 3 || liturgyDay!!.typeID == 4 || liturgyDay!!.typeID == 5) {
                sb.append(liturgyDay!!.breviaryHour!!.getIntermedia()!!.forRead)
            }
            if (liturgyDay!!.typeID == 6) {
                sb.append(liturgyDay!!.breviaryHour!!.getVisperas()!!.getAllForRead())
            }
            if (liturgyDay!!.typeID == 7) {
                sb.append(liturgyDay!!.breviaryHour!!.getCompletas()!!.getForRead())
            }
        } catch (e: Exception) {
            sb.append(Utils.createErrorMessage(e.message))
        }
        return sb
    }

    fun getSingleForRead(): StringBuilder {
        val sb = StringBuilder()
        try {
            sb.append(Utils.pointAtEnd(fecha))
            sb.append(Utils.pointAtEnd(tiempo))

            sb.append(tituloForRead)
        } catch (e: Exception) {
            sb.append(Utils.createErrorMessage(e.message))
        }
        return sb
    }
}